pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go

```diff
@@ -329,27 +317,40 @@ func (a *index) lookupService(key string) *model.ServiceInfo {
 
 // All return all known workloads. Result is un-ordered
 func (a *index) All() []model.AddressInfo {
+	res := dedupeWorkloads(a.workloads.List())
+	for _, s := range a.services.List() {
+		res = append(res, serviceToAddressInfo(s.Service))
+	}
+	return res
+}
+
+func dedupeWorkloads(workloads []model.WorkloadInfo) []model.AddressInfo {
+	if len(workloads) <= 1 {
+		return slices.Map(workloads, modelWorkloadToAddressInfo)
+	}
 	res := []model.AddressInfo{}
 	seenAddresses := sets.New[netip.Addr]()
-	for _, wl := range a.workloads.List() {
+	for _, wl := range workloads {
 		write := true
-		for _, addr := range wl.Addresses {
-			a := byteIPToAddr(addr)
-			if seenAddresses.InsertContains(a) {
-				// We have already seen this address. We don't want to include it.
-				// We do want to prefer Pods > WorkloadEntry to give precedence to Kubernetes. However, the underlying `a.workloads`
-				// already guarantees this, so no need to handle it here.
-				write = false
-				break
+		// HostNetwork mode is expected to have overlapping IPs, and tells the data plane to avoid relying on the IP as a unique
+		// identifier.
+		// For anything else, exclude duplicates.
+		if wl.NetworkMode != workloadapi.NetworkMode_HOST_NETWORK {
+			for _, addr := range wl.Addresses {
+				a := byteIPToAddr(addr)
+				if seenAddresses.InsertContains(a) {
+					// We have already seen this address. We don't want to include it.
+					// We do want to prefer Pods > WorkloadEntry to give precedence to Kubernetes. However, the underlying `a.workloads`
+					// already guarantees this, so no need to handle it here.
+					write = false
+					break
+				}
 			}
 		}
 		if write {
 			res = append(res, workloadToAddressInfo(wl.Workload))
 		}
 	}
-	for _, s := range a.services.List() {
-		res = append(res, serviceToAddressInfo(s.Service))
-	}
```

Comment:
this is just a weird diff, I didn't remove the service listing :slightly_smiling_face: